Define release checklist (#500)
kyma-project/test-infra: docs/prow/release-checklist.md

1. Create a release branch in `test-infra` repository.
This part applies only for new major and minor versions.
The name of this branch should follow the `release-x.y` pattern, such as `release-0.6`.

2. Ensure that the `prow/RELEASE_VERSION` file from `test-infra` repository on a release branch contains correct version to be created.
File should contains version if format `<A>.<B>.<C>` or `<A>.<B>.<C>-rc<D>` where `A`,`B`, `C` and `D` are numbers.
If you define release candidate version, pre-release will be created.

3. Create a release branch in the `kyma` repository. Do it only for a new release, not for a bugfix release.
The name of this branch should follow the `release-x.y` pattern, such as `release-0.6`.

4. Create a PR for the `kyma` release branch.

![](./assets/release-PR.png)
This triggers all jobs for components.
In a PR you have to update version and directory of used components in `values.yaml` files.

Change from:

```
dir: develop/
version: <current_version>
```
to:
```
dir:
version: <rel_version>
```    
    
Every component image is published with a version defined in the `RELEASE_VERSION` file stored in the `test-infra` repository on the given release branch. 

> 5. In case of a temporary failure of any job, you can retrigger it by adding a comment to the PR:

Reviewer: ```suggestion
5. If any job fails, retrigger it by adding the following comment to the PR:
```

```

> /test <job_name>

Reviewer: ```suggestion
/test {job_name}
```

```

> 6. Wait until all components for jobs or tools will be finished. 

Reviewer: ```suggestion
6. Wait until all jobs for components and tools finish. 
```

> `kyma-integration`, `kyma-gke-integration`, `kyma-artifacts` and `kyma-installer` needs to be executed manually because there

Reviewer: ```suggestion
Execute `kyma-integration`, `kyma-gke-integration`, `kyma-artifacts`, and `kyma-installer` manually because there
```

Author: Execution is described in the next steps. Here I want to only explain that you should not be surprised that some jobs are not triggered

> are dependencies between jobs. See diagram: 

Reviewer: ```suggestion
are dependencies between these jobs. See the diagram for details: 
```

![](./assets/kyma-rel-jobs.svg)

> 7.  Execute `kyma-integration` by adding comment on PR:

Reviewer: ```suggestion
7.  Execute `kyma-integration` by adding the `/test pre-rel06-kyma-integration` comment to the PR:
```

> `/test pre-rel06-kyma-integration`

Reviewer: ```suggestion
```

> 8. Execute one after the other `kyma-installer`, `kyma-artifacts` and `kyma-gke-integration` jobs. 

Reviewer: ```suggestion
8. Execute `kyma-installer`, `kyma-artifacts`, and `kyma-gke-integration` one after another. 
```

> You don't have to wait until `pre-rel06-kyma-integration` job will be finished.

Reviewer: ```suggestion
You don't have to wait until `pre-rel06-kyma-integration` job finishes.
```

> 5. In case of  detecting problems with a release, for example failing tests, you need to wait for the fix (fixed on PR or cherry picked to the PR). 

Reviewer: ```suggestion
5. If you detect any problems with the release, such as failing tests, wait for the fix that can be delivered either on a PR or cherry-picked to the PR from the `master` branch. 
```

> Then all jobs will be triggered one more time and you need to return to point 6.

Reviewer: ```suggestion
Prow triggers the jobs again. Return to point 6 to rerun manual jobs.
```

> 9. After all checks pass, merge PR.

Reviewer: ```suggestion
9. After all checks pass, merge the PR.
```

> 10. Merge to the release branch, executes Postsubmit job that creates a Github release.

Reviewer: ```suggestion
10. Merging the PR to the release branch executes the postsubmit job that creates a Github release.
```

> Validate generated `yamls` and changelog files: https://github.com/kyma-project/kyma/releases

Reviewer: ```suggestion
Validate the `yaml` and changelog files generated under [releases](https://github.com/kyma-project/kyma/releases).
```

> Update manually release content with instruction how to install released Kyma.

Reviewer: ```suggestion
Update the release content manually with the instruction on how to install the latest Kyma release.
```

> 11. Update `RELEASE_VERSION` to next version immediately after the release, both on master and release branch. If not, any PR to a release branch done by 

Reviewer: ```suggestion
11. Update `RELEASE_VERSION` to the next version both on the `master` and release branch. Do it immediately after the release, otherwise any PR to a release branch done by 
```

> Kyma developer overrides previously published Docker images.  

Reviewer: ```suggestion
a Kyma developer overrides the previously published Docker images.  
```

# Release checklist

This document describes how to execute release of Kyma using prow.

## Release preparation

This part can be done at any point of time. 
This part applies only for new major and minor versions.

1. Define release jobs in test-infra repository
 - for every component
 - for every tool

>  - kyma-integration

Reviewer 2: On list should be also `for every test` and there is one file in `kyma/docs`

 - kyma-gke-integration
 - kyma-artifacts
 - kyma-installer

To see all release jobs for release 0.6, look for job names with prefix: `pre-rel06`.
Every job name has to have unique name, so remember about prefixing it with `pre-rel<XY>`

> 2. Ensure that tests for jobs exist

Reviewer 2: At the same point should be something like: Add release branch name to `development/tools/jobs/tester/tester.go` file in `GetAllKymaReleaseBranches()` function.

3. Define branch protection rules for the release branch.

> For release-0.6 definition was following:

Reviewer 2: Maybe there should be information where add those rules? In `prow/config.yaml` file on key `branch-protection.orgs.kyma-project.kyma.branches`

```
release-0.6:
  protect: true
  required_status_checks:
    contexts:
      - pre-rel06-kyma-integration
      - pre-rel06-kyma-gke-integration
      - pre-rel06-kyma-artifacts
      - pre-rel06-kyma-installer
```

## Release

1. Create release branch in test-infra repository.
This part applies only for new major and minor versions.
The name of this branch should follow the `release-x.y` pattern, such as `release-0.6`.

> 2. Ensure that `RELEASE_VERSION` file contains version to be created.

Reviewer 2: Ensure that `prow/RELEASE_VERSION`...
and maybe add some examples like: `0.6.0-rc1`, `0.6.0` etc - it suggests how content file should looks like and also there should be every version no only major and minor

3. Create a release branch in the `kyma` repository. Do it only for a new release, not for a bugfix release.
The name of this branch should follow the `release-x.y` pattern, such as `release-0.6`.

4. Create a PR for the `kyma` release branch. This triggers all jobs for components.
In PR you have to update version and directory of used components in `values.yaml` files.

Change from:

```
dir: develop/
version: <current_version>
```
to:
```
dir:
version: <rel_version>
```    
    
Every component image is published with a version defined in the `RELEASE_VERSION` file stored in the `test-infra` repository on the given release branch. For example, it can include the `0.6.0-rc1` version.

5. In case of a temporary failure of any job, you can retrigger it by adding a comment:
```
test <job_name>

> ```

Reviewer 2: `/test <job_name>` it's how exactly comments looks like

6. Wait until all components for jobs or tools will be finished. 
`kyma-integration`, `kyma-gke-integration`, `kyma-artifacts` and `kyma-installer` needs to be executed manually, because there

> are dependencies between jobs. See diagram: ![](./assets/kyma-rel-jobs.svg)

Reviewer 2: add image to next line, on view it cuts me sentence, "See <image> diagram:"

7.  Execute `kyma-integration` by adding comment on PR:
`/test pre-rel06-kyma-integration`

> 8. Execute one after the other `kyma-installer`, `kyma-artifacts` and `kyma-gke-integration` jobs

Reviewer 2: Maybe: "Execute `kyma-installer` by adding comment on PR: `/test kyma-installer` when jobs will finish run next job `kyma-artifacts` ... .
This can prevent the situation when someone run jobs one after one but not wait until before job finished.
